Makefile

```diff
@@ -0,0 +1,111 @@
+# -*- coding: utf-8; mode: makefile-gmake -*-
+
+.DEFAULT = help
+DEBUG ?= --pdb
+TEST ?= .
+
+# python version to use
+PY       ?=3
+# list of python packages (folders) or modules (files) of this build
+PYOBJECTS = hawkauthlib
+# folder where the python distribution takes place
+PYDIST   ?= dist
+# folder where the python intermediate build files take place
+PYBUILD  ?= build
+
+SYSTEMPYTHON = `which python$(PY) python | head -n 1`
+VIRTUALENV   = virtualenv --python=$(SYSTEMPYTHON)
+VTENV_OPTS   = "--no-site-packages"
+TEST_FOLDER  = ./hawkauthlib/tests
+
+ENV     = ./local/py$(PY)
+ENV_BIN = $(ENV)/bin
+
+
+PHONY += help
+help::
+	@echo  'usage:'
+	@echo
+	@echo  '  build     - build virtualenv ($(ENV)) and install *developer mode*'
+	@echo  '  lint      - run pylint within "build" (developer mode)'
+	@echo  '  test      - run tests for all supported environments (tox)'
+	@echo  '  debug     - run tests within a PDB debug session'
+	@echo  '  dist      - build packages in "$(PYDIST)/"'
+	@echo  '  pypi      - upload "$(PYDIST)/*" files to PyPi'
+	@echo  '  clean	    - remove most generated files'
+	@echo
+	@echo  'options:'
+	@echo
+	@echo  '  PY=3      - python version to use (default 3)'
+	@echo  '  TEST=.    - choose test from $(TEST_FOLDER) (default "." runs all)'
+	@echo  '  DEBUG=    - target "debug": do not invoke PDB on errors'
+	@echo
+	@echo  'When using target "debug", set breakpoints within py-source by adding::'
+	@echo  '    ...'
+	@echo  '    DEBUG()'
+	@echo  '    ...'
+	@echo
+	@echo  'Example; a clean and fresh build (in local/py3), run all tests (py27, py35, lint)::'
+	@echo
+	@echo  '  make clean build test'
+	@echo
+	@echo  'Example; debug "test_<name>.py" within a python2 session (in local/py2)::'
+	@echo
+	@echo  '  make PY=2 TEST=test_<name>.py debug'
+	@echo
+
+
+PHONY += build
+build: $(ENV)
+	$(ENV_BIN)/pip -v install -e .
+
+
+PHONY += lint
+lint: $(ENV)
+	$(ENV_BIN)/pylint $(PYOBJECTS) --rcfile pylintrc
+
+PHONY += test
+test:  $(ENV)
+	$(ENV_BIN)/tox -vv
+
+# set breakpoint with:
+#    DEBUG()
+# e.g. to run tests in debug mode in emacs use:
+#   'M-x pdb' ... 'make debug'
+
+PHONY += debug
+debug: build
+	DEBUG=$(DEBUG) $(ENV_BIN)/pytest $(DEBUG) -v $(TEST_FOLDER)/$(TEST)
+
+$(ENV):
+	$(VIRTUALENV) $(VTENV_OPTS) $(ENV)
+	$(ENV_BIN)/pip install -r requirements.txt
+
+# for distribution, use python from virtualenv
+PHONY += dist
+dist:  clean-dist $(ENV)
+	$(ENV_BIN)/python setup.py \
+		sdist -d $(PYDIST)  \
+		bdist_wheel --bdist-dir $(PYBUILD) -d $(PYDIST)
+
+PHONY += pypi
+pypi: dist
+	$(ENV_BIN)/twine upload $(PYDIST)/*
```

Comment:
naming suggestion: `make publish` might be a bit better fit than `make pypi`?

Reply:
Since this is *publishing* at PyPi, I thought `pypi` is a god name for such a target. But it is up to you, if you want, we can name the target `publish`.

Reply:
I mildly prefer `make publish`, if you're happy then let's go with that.